pkg/dataobj/consumer/partition_processor.go

```diff
@@ -177,7 +193,19 @@ func (p *partitionProcessor) processRecord(record *kgo.Record) {
 			backoff.Wait()
 		}
 
-		if err := p.metastoreManager.UpdateMetastore(p.ctx, flushResult); err != nil {
+		backoff.Reset()
+		var objectPath string
+		for backoff.Ongoing() {
+			objectPath, err = p.uploader.Upload(p.ctx, p.flushBuffer)
+			if err == nil {
+				break
+			}
+			level.Error(p.logger).Log("msg", "failed to upload object", "err", err)
+			p.metrics.incUploadFailures()
+			backoff.Wait()
+		}
+
+		if err := p.metastoreManager.UpdateMetastore(p.ctx, objectPath, flushedDataobjStats); err != nil {
```

Comment:
subtle edge case here: if the p.uploader.Upload never succeeds in the previous backoff loop, we always update the metastore anyway, pointing to an object that didn't get uploaded (ie we're missing a check for `err != nil` or `backoff.Err() != nil` after the loop)